manpages/rebar3.1
.TH "REBAR3" "1" "November 2018" "Erlang"

.SH NAME

\fBrebar3\fR \- tool for working with Erlang projects

.SH "SYNOPSIS"

\fBrebar3\fR \fB\-\-version\fR

.br
\fBrebar3\fR \fBhelp\fR

.br
\fBrebar3\fR \fIcommand\fR [\fIoptions\fR] \.\.\.

.SH "DESCRIPTION"

Rebar3 is an Erlang tool that makes it easy to create, develop, and release Erlang libraries, applications, and systems in a repeatable manner\.

Full documentation at \fIhttp://www.rebar3.org/\fR

.SH "ESSENTIAL COMMANDS"

For the full command list, see the \fICOMMANDS\fR section\.

.P
With \fBrebar3 help <task>\fR, commands and plugins will display their own help information\.

.TP
\fBcompile\fR
Compile the current project

.TP
\fBnew\fR (\fBhelp [\fItemplate\fR])|\fItemplate\fR
Show information about templates or use one

.TP
\fBupdate\fR
Fetch the newest version of the Hex index

.TP
\fBdo\fR \fIcommand\fR[,\fIcommand\fR,...]
Run one or more commands in a sequence

.TP
\fBshell\fR
Start the current project in a REPL\. You can then use \fBr3:do(\fItask\fR)\fR to call it on the project without dropping the current state.

.SH "COMMANDS"

. this section generated by running the contents of 'commands' in rebar3 shell

.TP
\fBas\fR 
Higher order provider for running multiple tasks in a sequence as a certain profiles.
.TP
\fBclean\fR [\fI-a\fR|\fI--all\fR] [\fI-p\fR|\fI--profile\fR] 
Remove compiled beam files from apps.
.IP
\fI--all\fR: Clean all apps include deps
.IP
\fI--profile\fR: Clean under profile. Equivalent to `rebar3 as <profile> clean`
.TP
\fBcompile\fR [\fI-d\fR|\fI--deps_only\fR] 
Compile apps .app.src and .erl files.
.IP
\fI--deps_only\fR: Only compile dependencies, no project apps will be built.
.TP
\fBcover\fR [\fI-r\fR|\fI--reset\fR] [\fI-v\fR|\fI--verbose\fR] [\fI-m\fR|\fI--min_coverage\fR] 
Perform coverage analysis.
.IP
\fI--reset\fR: Reset all coverdata.
.IP
\fI--verbose\fR: Print coverage analysis.
.IP
\fI--min_coverage\fR: Mandate a coverage percentage required to succeed (0..100)
.TP
\fBct\fR [\fI--dir\fR] [\fI--suite\fR] [\fI--group\fR] [\fI--case\fR] [\fI--label\fR] [\fI--config\fR] [\fI--spec\fR] [\fI--join_specs\fR] [\fI--allow_user_terms\fR] [\fI--logdir\fR] [\fI--logopts\fR] [\fI--verbosity\fR] [\fI-c\fR|\fI--cover\fR] [\fI--cover_export_name\fR] [\fI--repeat\fR] [\fI--duration\fR] [\fI--until\fR] [\fI--force_stop\fR] [\fI--basic_html\fR] [\fI--stylesheet\fR] [\fI--decrypt_key\fR] [\fI--decrypt_file\fR] [\fI--abort_if_missing_suites\fR] [\fI--multiply_timetraps\fR] [\fI--scale_timetraps\fR] [\fI--create_priv_dir\fR] [\fI--include\fR] [\fI--readable\fR] [\fI-v\fR|\fI--verbose\fR] [\fI--name\fR] [\fI--sname\fR] [\fI--setcookie\fR] [\fI--sys_config\fR] [\fI--compile_only\fR] [\fI--retry\fR] 
Run Common Tests.
.IP
\fI--dir\fR: List of additional directories containing test suites
.IP
\fI--suite\fR: List of test suites to run
.IP
\fI--group\fR: List of test groups to run
.IP
\fI--case\fR: List of test cases to run
.IP
\fI--label\fR: Test label
.IP
\fI--config\fR: List of config files
.IP
\fI--spec\fR: List of test specifications
.IP
\fI--join_specs\fR: Merge all test specifications and perform a single test run
.IP
\fI--allow_user_terms\fR: Allow user defined config values in config files
.IP
\fI--logdir\fR: Log folder
.IP
\fI--logopts\fR: Options for common test logging
.IP
\fI--verbosity\fR: Verbosity
.IP
\fI--cover\fR: Generate cover data
.IP
\fI--cover_export_name\fR: Base name of the coverdata file to write
.IP
\fI--repeat\fR: How often to repeat tests
.IP
\fI--duration\fR: Max runtime (format: HHMMSS)
.IP
\fI--until\fR: Run until (format: HHMMSS)
.IP
\fI--force_stop\fR: Force stop on test timeout (true | false | skip_rest)
.IP
\fI--basic_html\fR: Show basic HTML
.IP
\fI--stylesheet\fR: CSS stylesheet to apply to html output
.IP
\fI--decrypt_key\fR: Path to key for decrypting config
.IP
\fI--decrypt_file\fR: Path to file containing key for decrypting config
.IP
\fI--abort_if_missing_suites\fR: Abort if suites are missing
.IP
\fI--multiply_timetraps\fR: 
.IP
\fI--scale_timetraps\fR: Scale timetraps
.IP
\fI--create_priv_dir\fR: Create priv dir (auto_per_run | auto_per_tc | manual_per_tc)
.IP
\fI--include\fR: Directories containing additional include files
.IP
\fI--readable\fR: Shows test case names and only displays logs to shell on failures (true | compact | false)
.IP
\fI--verbose\fR: Verbose output
.IP
\fI--name\fR: Gives a long name to the node
.IP
\fI--sname\fR: Gives a short name to the node
.IP
\fI--setcookie\fR: Sets the cookie if the node is distributed
.IP
\fI--sys_config\fR: List of application config files
.IP
\fI--compile_only\fR: Compile modules in the project with the test configuration but do not run the tests
.IP
\fI--retry\fR: Experimental feature. If any specification for previously failing test is found, runs them.
.TP
\fBdeps\fR 
List dependencies
.TP
\fBdialyzer\fR [\fI-u\fR|\fI--update-plt\fR] [\fI-s\fR|\fI--succ-typings\fR] 
Run the Dialyzer analyzer on the project.
.IP
\fI--update-plt\fR: Enable updating the PLT. Default: true
.IP
\fI--succ-typings\fR: Enable success typing analysis. Default: true
.TP
\fBdo\fR 
Higher order provider for running multiple tasks in a sequence.
.TP
\fBedoc\fR 
Generate documentation using edoc.
.TP
\fBescriptize\fR 
Generate escript archive.
.TP
\fBeunit\fR [\fI--app\fR] [\fI--application\fR] [\fI-c\fR|\fI--cover\fR] [\fI--cover_export_name\fR] [\fI-d\fR|\fI--dir\fR] [\fI-f\fR|\fI--file\fR] [\fI-m\fR|\fI--module\fR] [\fI-s\fR|\fI--suite\fR] [\fI-v\fR|\fI--verbose\fR] [\fI--name\fR] [\fI--sname\fR] [\fI--setcookie\fR] 
Run EUnit Tests.
.IP
\fI--app\fR: Comma separated list of application test suites to run. Equivalent to `[{application, App}]`.
.IP
\fI--application\fR: Comma separated list of application test suites to run. Equivalent to `[{application, App}]`.
.IP
\fI--cover\fR: Generate cover data. Defaults to false.
.IP
\fI--cover_export_name\fR: Base name of the coverdata file to write
.IP
\fI--dir\fR: Comma separated list of dirs to load tests from. Equivalent to `[{dir, Dir}]`.
.IP
\fI--file\fR: Comma separated list of files to load tests from. Equivalent to `[{file, File}]`.
.IP
\fI--module\fR: Comma separated list of modules to load tests from. Equivalent to `[{module, Module}]`.
.IP
\fI--suite\fR: Comma separated list of modules to load tests from. Equivalent to `[{module, Module}]`.
.IP
\fI--verbose\fR: Verbose output. Defaults to false.
.IP
\fI--name\fR: Gives a long name to the node
.IP
\fI--sname\fR: Gives a short name to the node
.IP
\fI--setcookie\fR: Sets the cookie if the node is distributed
.TP
\fBget-deps\fR 
Fetch dependencies.
.TP
\fBhelp\fR 
Display a list of tasks or help for a given task or subtask.
.TP
\fBnew\fR [\fI-f\fR|\fI--force\fR] 
Create new project from templates.
.IP
\fI--force\fR: overwrite existing files
.TP
\fBpath\fR [\fI--app\fR] [\fI--base\fR] [\fI--bin\fR] [\fI--ebin\fR] [\fI--lib\fR] [\fI--priv\fR] [\fI-s\fR|\fI--separator\fR] [\fI--src\fR] [\fI--rel\fR] 
Print paths to build dirs in current profile.
.IP
\fI--app\fR: Comma separated list of applications to return paths for.
.IP
\fI--base\fR: Return the `base' path of the current profile.
.IP
\fI--bin\fR: Return the `bin' path of the current profile.
.IP
\fI--ebin\fR: Return all `ebin' paths of the current profile's applications.
.IP
\fI--lib\fR: Return the `lib' path of the current profile.
.IP
\fI--priv\fR: Return the `priv' path of the current profile's applications.
.IP
\fI--separator\fR: In case of multiple return paths, the separator character to use to join them.
.IP
\fI--src\fR: Return the `src' path of the current profile's applications.
.IP
\fI--rel\fR: Return the `rel' path of the current profile.
.TP
\fBpkgs\fR 
List information for a package.
.TP
\fBrelease\fR [\fI-n\fR|\fI--relname\fR] [\fI-v\fR|\fI--relvsn\fR] [\fI-g\fR|\fI--goal\fR] [\fI-u\fR|\fI--upfrom\fR] [\fI-o\fR|\fI--output-dir\fR] [\fI-h\fR|\fI--help\fR] [\fI-l\fR|\fI--lib-dir\fR] [\fI-p\fR|\fI--path\fR] [\fI--default-libs\fR] [\fI-V\fR|\fI--verbose\fR] [\fI-d\fR|\fI--dev-mode\fR] [\fI-i\fR|\fI--include-erts\fR] [\fI-a\fR|\fI--override\fR] [\fI-c\fR|\fI--config\fR] [\fI--overlay_vars\fR] [\fI--vm_args\fR] [\fI--sys_config\fR] [\fI--system_libs\fR] [\fI--version\fR] [\fI-r\fR|\fI--root\fR] 
Build release of project.
.IP
\fI--relname\fR: Specify the name for the release that will be generated
.IP
\fI--relvsn\fR: Specify the version for the release
.IP
\fI--goal\fR: Specify a target constraint on the system. These are usually the OTP
.IP
\fI--upfrom\fR: Only valid with relup target, specify the release to upgrade from
.IP
\fI--output-dir\fR: The output directory for the release. This is `./` by default.
.IP
\fI--help\fR: Print usage
.IP
\fI--lib-dir\fR: Additional dir that should be searched for OTP Apps
.IP
\fI--path\fR: Additional dir to add to the code path
.IP
\fI--default-libs\fR: Whether to use the default system added lib dirs (means you must add them all manually). Default is true
.IP
\fI--verbose\fR: Verbosity level, maybe between 0 and 3
.IP
\fI--dev-mode\fR: Symlink the applications and configuration into the release instead of copying
.IP
\fI--include-erts\fR: If true include a copy of erts used to build with, if a path include erts at that path. If false, do not include erts
.IP
\fI--override\fR: Provide an app name and a directory to override in the form <appname>:<app directory>
.IP
\fI--config\fR: The path to a config file
.IP
\fI--overlay_vars\fR: Path to a file of overlay variables
.IP
\fI--vm_args\fR: Path to a file to use for vm.args
.IP
\fI--sys_config\fR: Path to a file to use for sys.config
.IP
\fI--system_libs\fR: Path to dir of Erlang system libs
.IP
\fI--version\fR: Print relx version
.IP
\fI--root\fR: The project root directory
.TP

RELEASE-NOTES.md
# Rebar 2.5.0

* Reverted rebar/281: [Move include/rebar.hrl to src/rebar.hrl](https://github.com/rebar/rebar/pull/281) as it broke backwards compatibility

# Rebar 2.4.0

* rebar/52: [Slim release support](https://github.com/rebar/rebar/pull/52)
* rebar/112: [Add code coverage analysis functionality to `qc'](https://github.com/rebar/rebar/pull/112)
* rebar/119: [Add qualified name tests specification (see #118)](https://github.com/rebar/rebar/pull/119)
* rebar/130: [ct fixes](https://github.com/rebar/rebar/pull/130)
* rebar/136: [Add support for the Perforce VCS client via the "p4" tool](https://github.com/rebar/rebar/pull/136)
* rebar/195: [Switch template instructions](https://github.com/rebar/rebar/pull/195)
* rebar/229: [Add REBAR to environment before executing hooks](https://github.com/rebar/rebar/pull/229)
* rebar/260: [Quote include/lib paths to handle spaces in Erlang installs (fixes build on windows)](https://github.com/rebar/rebar/pull/260)
* rebar/280: [improve output when using `rebar shell`](https://github.com/rebar/rebar/pull/280)
* rebar/281: [Move include/rebar.hrl to src/rebar.hrl](https://github.com/rebar/rebar/pull/281)
* rebar/284: [Error 'Command not found' when sname is used](https://github.com/rebar/rebar/pull/284)
* rebar/285: [Fix #249 (erlc regression)](https://github.com/rebar/rebar/pull/285)
* rebar/288: [Extend and document contributing rules](https://github.com/rebar/rebar/pull/288)
* rebar/289: [erlc: fix typo in update_erlcinfo/3 clause that would make the function fail](https://github.com/rebar/rebar/pull/289)
* rebar/290: [erlc: replace if expression with case of](https://github.com/rebar/rebar/pull/290)
* rebar/292: [Namespaced types: fix build for 17.0](https://github.com/rebar/rebar/pull/292)
* rebar/296: [Add gen_event template](https://github.com/rebar/rebar/pull/296)


# Rebar 2.3.1

## PR's Merged

* rebar/242: [Extra commits for #129](https://github.com/rebar/rebar/pull/242)
* rebar/244: [Document skip_apps=, apps=, and require_*_vsn](https://github.com/rebar/rebar/pull/244)
* rebar/251: [Make sure that eunit/qc_compile_opts works](https://github.com/rebar/rebar/pull/251)
* rebar/255: [rebar.app: remove superfluous quoting](https://github.com/rebar/rebar/pull/255)
* rebar/274: [Use lowercase for Windows drive name to resolve issue #250](https://github.com/rebar/rebar/pull/274)

# Rebar 2.3.0

## PR's Merged

* rebar/98: [Repetition of environment variable definitions in child processes (ports)](https://github.com/rebar/rebar/pull/98)
* rebar/115: [Incorrect REMSH args when sname is used.](https://github.com/rebar/rebar/pull/115)
* rebar/129: [Speed up the compilation process v5](https://github.com/rebar/rebar/pull/129)
* rebar/139: [Allow specification of module dependencies for appups](https://github.com/rebar/rebar/pull/139)
* rebar/175: [CWD plugins regression](https://github.com/rebar/rebar/pull/175)
* rebar/188: [Xref extra path](https://github.com/rebar/rebar/pull/188)
* rebar/208: [Fix typo in rebar_erlydtl_compiler](https://github.com/rebar/rebar/pull/208)
* rebar/219: [Added R16B01 and R16B02 to travis config.](https://github.com/rebar/rebar/pull/219)
* rebar/221: [Adapt erlydtl compiler plugin to latest version of erlydtl](https://github.com/rebar/rebar/pull/221)
* rebar/223: [Add random_suite_order option to eunit command](https://github.com/rebar/rebar/pull/223)
* rebar/224: [allow suites or tests as options for eunit and ct](https://github.com/rebar/rebar/pull/224)
* rebar/230: [eunit: fix dialyzer warnings introduced in 03da5e0b](https://github.com/rebar/rebar/pull/230)
* rebar/232: [Document support for abbreviated commands](https://github.com/rebar/rebar/pull/232)
* rebar/233: [docs: fix #228](https://github.com/rebar/rebar/pull/233)
* rebar/234: [Fix #220 (Reported-by: Joseph Norton)](https://github.com/rebar/rebar/pull/234)
* rebar/237: [Add partial support for Erlang/OTP 17](https://github.com/rebar/rebar/pull/237)
* rebar/252: [file_utils: properly report errors (fix #95)](https://github.com/rebar/rebar/pull/252)
* rebar/254: [Fix 'rebar generate' regression (#253)](https://github.com/rebar/rebar/pull/254)
* rebar/265: [Fix 'rebar help clean' function_clause error](https://github.com/rebar/rebar/pull/265)
* rebar/268: [Fix #267 (code path regression)](https://github.com/rebar/rebar/pull/268)
* rebar/269: [Update THANKS](https://github.com/rebar/rebar/pull/269)

# Rebar 2.2.0

## PR's Merged

* rebar/152: [Fix erl_opts use](https://github.com/rebar/rebar/pull/152)
* rebar/154: [Fix update-deps with certain forms of the {tag, ...} type](https://github.com/rebar/rebar/pull/154)
* rebar/155: [Fixes for #137 and #142](https://github.com/rebar/rebar/pull/155)

src/rebar3.erl
%% -*- erlang-indent-level: 4;indent-tabs-mode: nil -*-
%% ex: ts=4 sw=4 et
%% -------------------------------------------------------------------
%%
%% rebar: Erlang Build Tools
%%
%% Copyright (c) 2009 Dave Smith (dizzyd@dizzyd.com)
%%
%% Permission is hereby granted, free of charge, to any person obtaining a copy
%% of this software and associated documentation files (the "Software"), to deal
%% in the Software without restriction, including without limitation the rights
%% to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
%% copies of the Software, and to permit persons to whom the Software is
%% furnished to do so, subject to the following conditions:
%%
%% The above copyright notice and this permission notice shall be included in
%% all copies or substantial portions of the Software.
%%
%% THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
%% IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
%% FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
%% AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
%% LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
%% OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
%% THE SOFTWARE.
%% -------------------------------------------------------------------
%%
%% @doc Main module for rebar3. Supports two interfaces; one for escripts,
%% and one for usage as a library (although rebar3 makes a lot of
%% assumptions about its environment, making it a bit tricky to use as
%% a lib).
%%
%% This module's job is mostly to set up the root environment for rebar3
%% and handle global options (mostly all from the ENV) and make them
%% accessible to the rest of the run.
%% @end
-module(rebar3).

-export([main/0,
         main/1,
         run/1,
         run/2,
         global_option_spec_list/0,
         init_config/0,
         set_options/2,
         parse_args/1,
         version/0,
         log_level/0]).

-include("rebar.hrl").

%% ====================================================================
%% Public API
%% ====================================================================

%% @doc For running with:
%% erl +sbtu +A1 -noinput -mode minimal -boot start_clean -s rebar3 main -extra "$@"
-spec main() -> no_return().
main() ->
    List = init:get_plain_arguments(),
    main(List).

%% @doc escript Entry point
-spec main(list()) -> no_return().
main(Args) ->
    try run(Args) of
        {ok, _State} ->
            erlang:halt(0);
        Error ->
            handle_error(Error, [])
    catch
        ?WITH_STACKTRACE(_,Error,Stacktrace)
            handle_error(Error, Stacktrace)
    end.

%% @doc Erlang-API entry point
-spec run(rebar_state:t(), [string()]) -> {ok, rebar_state:t()} | {error, term()}.
run(BaseState, Commands) ->
    start_and_load_apps(api),
    BaseState1 = rebar_state:set(BaseState, task, Commands),
    BaseState2 = rebar_state:set(BaseState1, caller, api),

    Verbosity = log_level(),
    ok = rebar_log:init(api, Verbosity),

    run_aux(BaseState2, Commands).

%% ====================================================================
%% Internal functions
%% ====================================================================

%% @private sets up the rebar3 environment based on the command line
%% arguments passed, if they have any relevance; used to translate
%% from the escript call-site into a common one with the library
%% usage.
-spec run([any(), ...]) -> {ok, rebar_state:t()} | {error, term()}.
run(RawArgs) ->
    start_and_load_apps(command_line),

    BaseState = init_config(),
    BaseState1 = rebar_state:set(BaseState, caller, command_line),

    case erlang:system_info(version) of
        "6.1" ->
            ?WARN("Due to a filelib bug in Erlang 17.1 it is recommended"
                  "you update to a newer release.", []);
        _ ->
            ok
    end,

    {BaseState2, _Args1} = set_options(BaseState1, {[], []}),
    run_aux(BaseState2, RawArgs).

%% @private Junction point between the CLI and library entry points.
%% From here on the module's role is a shared path here to finish
%% up setting the environment for the run.
-spec run_aux(rebar_state:t(), [string()]) ->
    {ok, rebar_state:t()} | {error, term()}.
run_aux(State, RawArgs) ->
    io:setopts([{encoding, unicode}]),
    %% Profile override; can only support one profile
    State1 = case os:getenv("REBAR_PROFILE") of
                 false ->
                     State;
                 "" ->
                     State;
                 Profile ->
                     rebar_state:apply_profiles(State, [list_to_atom(Profile)])
             end,

    rebar_utils:check_min_otp_version(rebar_state:get(State1, minimum_otp_vsn, undefined)),
    rebar_utils:check_blacklisted_otp_versions(rebar_state:get(State1, blacklisted_otp_vsns, undefined)),

    %% Change the default hex CDN
    State2 = case os:getenv("HEX_CDN") of
                 false ->
                     State1;
                 CDN ->
                     rebar_state:set(State1, rebar_packages_cdn, CDN)
             end,

    Compilers = application:get_env(rebar, compilers, []),
    State0 = rebar_state:compilers(State2, Compilers),

    %% TODO: this means use of REBAR_PROFILE=profile will replace the repos with
    %% the repos defined in the profile. But it will not work with `as profile`.
    %% Maybe it shouldn't work with either to be consistent?
    Resources = application:get_env(rebar, resources, []),
    State2_ = rebar_state:create_resources(Resources, State0),

    %% bootstrap test profile
    State3 = rebar_state:add_to_profile(State2_, test, test_state(State1)),

    %% Process each command, resetting any state between each one
    BaseDir = rebar_state:get(State, base_dir, ?DEFAULT_BASE_DIR),
    State4 = rebar_state:set(State3, base_dir,
                             filename:join(filename:absname(rebar_state:dir(State3)), BaseDir)),

    {ok, Providers} = application:get_env(rebar, providers),
    %% Providers can modify profiles stored in opts, so set default after initializing providers
    State5 = rebar_state:create_logic_providers(Providers, State4),
    %% Initializing project_plugins which can override default providers
    State6 = rebar_plugins:project_plugins_install(State5),
    State7 = rebar_plugins:top_level_install(State6),
    State8 = case os:getenv("REBAR_CACHE_DIR") of
                false ->
                    State7;
                ConfigFile ->
                    rebar_state:set(State7, global_rebar_dir, ConfigFile)
            end,

    State9 = rebar_state:default(State8, rebar_state:opts(State8)),

    {Task, Args} = parse_args(RawArgs),

    State10 = rebar_state:code_paths(State9, default, code:get_path()),

    case rebar_core:init_command(rebar_state:command_args(State10, Args), Task) of
        {ok, State11} ->
            case rebar_state:get(State11, caller, command_line) of
                api ->
                    rebar_paths:unset_paths([deps, plugins], State11),
                    {ok, State11};
                _ ->
                    {ok, State11}
            end;
        Other ->
            Other
    end.



%% @doc set up base configuration having to do with verbosity, where
%% to find config files, and so on, and return an internal rebar3 state term.
-spec init_config() -> rebar_state:t().
init_config() ->
    rebar_utils:set_httpc_options(),

    %% Initialize logging system
    Verbosity = log_level(),
    ok = rebar_log:init(command_line, Verbosity),

    Config = rebar_config:consult_root(),
    Config1 = rebar_config:merge_locks(Config, rebar_config:consult_lock_file(?LOCK_FILE)),

    %% If $HOME/.config/rebar3/rebar.config exists load and use as global config
    GlobalConfigFile = rebar_dir:global_config(),
    State = case filelib:is_regular(GlobalConfigFile) of
                true ->
                    ?DEBUG("Load global config file ~ts", [GlobalConfigFile]),
                    try state_from_global_config(Config1, GlobalConfigFile)
                    catch
                        _:_ ->
                            ?WARN("Global config ~ts exists but can not be read. Ignoring global config values.", [GlobalConfigFile]),
                            rebar_state:new(Config1)
                    end;
                false ->
                    rebar_state:new(Config1)
            end,

    %% Determine the location of the rebar executable; important for pulling
    %% resources out of the escript
    State1 = try
                 ScriptName = filename:absname(escript:script_name()),
                 %% Running with 'erl -s rebar3 main' still sets a name for some reason
                 %% so verify it is a real file
                 case filelib:is_regular(ScriptName) of
                     true ->
                         rebar_state:escript_path(State, ScriptName);
                     false ->
                         State
                 end
             catch
                 _:_ ->
                     State
             end,

    %% TODO: Do we need this still? I think it may still be used.
    %% Initialize vsn cache
    rebar_state:set(State1, vsn_cache, dict:new()).

%% @doc Parse basic rebar3 arguments to find the top-level task
%% to be run; this parsing is only partial from the point of view that
%% runs done with arguments like `as $PROFILE do $TASK' will just
%% return `as', which is then in charge of doing a more dynamic
%% dispatch.
%% If no arguments are given, the `help' task is returned.
%% If special arguments like `-h' or `-v' are translated to `help'
%% and `version' tasks.
%% The unparsed parts of arguments are returned in:
%% `{Task, Rest}'.
-spec parse_args([string()]) -> {atom(), [string()]}.
parse_args([]) ->
    parse_args(["help"]);
parse_args([H | Rest]) when H =:= "-h"
                          ; H =:= "--help" ->
    parse_args(["help" | Rest]);
parse_args([H | Rest]) when H =:= "-v"
                          ; H =:= "--version" ->
    parse_args(["version" | Rest]);
parse_args([Task | RawRest]) ->
    {list_to_atom(Task), RawRest}.

%% @private actually not too sure what this does anymore.
-spec set_options(rebar_state:t(),{[any()],[any()]}) -> {rebar_state:t(),[any()]}.
set_options(State, {Options, NonOptArgs}) ->
    GlobalDefines = proplists:get_all_values(defines, Options),

    State1 = rebar_state:set(State, defines, GlobalDefines),

    %% Set global variables based on getopt options
    State2 = set_global_flag(State1, Options, force),

    Task = proplists:get_value(task, Options, "help"),

    {rebar_state:set(State2, task, Task), NonOptArgs}.

%% @doc get log level based on getopt options and ENV
-spec log_level() -> integer().
log_level() ->
    case os:getenv("QUIET") of
        Q when Q == false; Q == "" ->
            DefaultLevel = rebar_log:default_level(),
            case os:getenv("DEBUG") of
                D when D == false; D == "" ->
                    DefaultLevel;
                _ ->
                    DefaultLevel + 3
            end;
         _ ->
            rebar_log:error_level()
    end.

%% @doc show version information
-spec version() -> ok.
version() ->
    {ok, Vsn} = application:get_key(rebar, vsn),
    ?CONSOLE("rebar ~ts on Erlang/OTP ~ts Erts ~ts",
             [Vsn, erlang:system_info(otp_release), erlang:system_info(version)]).

%% @private set global flag based on getopt option boolean value
%% TODO: Actually make it 'global'
-spec set_global_flag(rebar_state:t(), list(), term()) -> rebar_state:t().
set_global_flag(State, Options, Flag) ->
    Value = case proplists:get_bool(Flag, Options) of
                true ->
                    "1";
                false ->
                    "0"
            end,
    rebar_state:set(State, Flag, Value).


%% @doc options accepted via getopt
-spec global_option_spec_list() -> [{atom(), char(), string(), atom(), string()}, ...].
global_option_spec_list() ->
    [
    %% {Name,  ShortOpt,  LongOpt,    ArgSpec,   HelpMsg}
    {help,     $h,        "help",     undefined, "Print this help."},
    {version,  $v,        "version",  undefined, "Show version information."},
    {task,     undefined, undefined,  string,    "Task to run."}
    ].

%% @private translate unhandled errors and internal return codes into proper
%% erroneous program exits.
-spec handle_error(term(), term()) -> no_return().
handle_error(rebar_abort, _) ->
    erlang:halt(1);
handle_error({error, rebar_abort}, _) ->
    erlang:halt(1);
handle_error({error, {Module, Reason}}, Stacktrace) ->
    case code:which(Module) of
        non_existing ->
            ?CRASHDUMP("~p: ~p~n~p~n~n", [Module, Reason, Stacktrace]),
            ?ERROR("Uncaught error in rebar_core. Run with DEBUG=1 to stacktrace or consult rebar3.crashdump", []),
            ?DEBUG("Uncaught error: ~p ~p", [Module, Reason]),
            ?INFO("When submitting a bug report, please include the output of `rebar3 report \"your command\"`", []);
        _ ->
            ?ERROR("~ts", [Module:format_error(Reason)])
    end,
    erlang:halt(1);
handle_error({error, Error}, _) when is_list(Error) ->
    ?ERROR("~ts", [Error]),
    erlang:halt(1);
handle_error(Error, StackTrace) ->
    %% Nothing should percolate up from rebar_core;
    %% Dump this error to console
    ?CRASHDUMP("Error: ~p~n~p~n~n", [Error, StackTrace]),
    ?ERROR("Uncaught error in rebar_core. Run with DEBUG=1 to see stacktrace or consult rebar3.crashdump", []),
    ?DEBUG("Uncaught error: ~p", [Error]),
    case StackTrace of
        [] -> ok;
        Trace ->
            ?DEBUG("Stack trace to the error location:~n~p", [Trace])
    end,
    ?INFO("When submitting a bug report, please include the output of `rebar3 report \"your command\"`", []),
    erlang:halt(1).

%% @private Boot Erlang dependencies; problem is that escripts don't auto-boot
%% stuff the way releases do and we have to do it by hand.
%% This also lets us detect and show nicer errors when a critical lib is
%% not supported
-spec start_and_load_apps(command_line|api) -> term().
start_and_load_apps(Caller) ->
    _ = application:load(rebar),
    %% Make sure crypto is running
    ensure_running(crypto, Caller),
    ensure_running(asn1, Caller),
    ensure_running(public_key, Caller),
    ensure_running(ssl, Caller),
    ensure_running(inets, Caller),
    inets:start(httpc, [{profile, rebar}]).

%% @doc Make sure a required app is running, or display an error message
%% and abort if there's a problem.
-spec ensure_running(atom(), command_line|api) -> ok | no_return().
ensure_running(App, Caller) ->
    case application:start(App) of
        ok -> ok;
        {error, {already_started, App}} -> ok;
        {error, Reason} ->
            %% These errors keep rebar3's own configuration to be loaded,
            %% which disables the log level and causes a failure without
            %% showing the error message. Bypass this entirely by overriding
            %% the default value (which allows logging to take place)
            %% and shut things down manually.
            Log = ec_cmd_log:new(warn, Caller),
            ec_cmd_log:error(Log, "Rebar dependency ~p could not be loaded "
                                  "for reason ~p~n", [App, Reason]),
            throw(rebar_abort)
    end.

-spec state_from_global_config([term()], file:filename()) -> rebar_state:t().
state_from_global_config(Config, GlobalConfigFile) ->
    GlobalConfigTerms = rebar_config:consult_file(GlobalConfigFile),
    GlobalConfig = rebar_state:new(GlobalConfigTerms),

    %% We don't want to worry about global plugin install state effecting later
    %% usage. So we throw away the global profile state used for plugin install.
    GlobalConfigThrowAway0 = rebar_state:current_profiles(GlobalConfig, [global]),

    Resources = application:get_env(rebar, resources, []),
    GlobalConfigThrowAway = rebar_state:create_resources(Resources, GlobalConfigThrowAway0),

    Compilers = application:get_env(rebar, compilers, []),
    GlobalConfigThrowAway1 = rebar_state:compilers(GlobalConfigThrowAway, Compilers),

    GlobalState = case rebar_state:get(GlobalConfigThrowAway1, plugins, []) of
                      [] ->
                          GlobalConfigThrowAway1;
                      GlobalPluginsToInstall ->
                          rebar_plugins:handle_plugins(global,
                                                       GlobalPluginsToInstall,
                                                       GlobalConfigThrowAway1)
                  end,
    GlobalPlugins = rebar_state:providers(GlobalState),
    GlobalConfig2 = rebar_state:set(GlobalConfig, plugins, []),
    GlobalConfig3 = rebar_state:set(GlobalConfig2, {plugins, global},
                                    rebar_state:get(GlobalConfigThrowAway1, plugins, [])),
    rebar_state:providers(rebar_state:new(GlobalConfig3, Config), GlobalPlugins).

-spec test_state(rebar_state:t()) -> [{'extra_src_dirs',[string()]} | {'erl_opts',[any()]}].
test_state(State) ->
    %% Fetch the test profile's erl_opts only
    Opts = rebar_state:opts(State),
    Profiles = rebar_opts:get(Opts, profiles, []),
    ProfileOpts = proplists:get_value(test, Profiles, []),
    ErlOpts = proplists:get_value(erl_opts, ProfileOpts, []),
    TestOpts = safe_define_test_macro(ErlOpts),
    [{extra_src_dirs, ["test"]}, {erl_opts, TestOpts}].

-spec safe_define_test_macro([any()]) -> [any()] | [{'d',atom()} | any()].
safe_define_test_macro(Opts) ->
    %% defining a compile macro twice results in an exception so
    %% make sure 'TEST' is only defined once
    case test_defined(Opts) of
       true  -> Opts;
       false -> [{d, 'TEST'}|Opts]
    end.

-spec test_defined([{d, atom()} | {d, atom(), term()} | term()]) -> boolean().
test_defined([{d, 'TEST'}|_]) -> true;
test_defined([{d, 'TEST', true}|_]) -> true;
test_defined([_|Rest]) -> test_defined(Rest);
test_defined([]) -> false.
